File: firebase/functions/lib/utils/infra/firebase_auth_utils.dart
import './firestore_utils.dart';
import 'package:firebase_admin_interop/firebase_admin_interop.dart';

/// Use a global singleton to allow mocking in tests
FirebaseAuthUtils firebaseAuthUtils = FirebaseAuthUtils();

class FirebaseAuthUtils {
  Future<UserRecord> getUser(String uid) async {
    // Get user record
    var user = await firebaseApp.auth().getUser(uid);

    return user;
  }

  Future<List<UserRecord>> getUsers(List<String> uids) async {
    // Get emails of every user we want to email
    final getUserFutures = uids.map((id) => firebaseApp.auth().getUser(id));

    return Future.wait(getUserFutures);
  }

  Future<UserRecord?> getUserByEmailIfExists(String email) async {
    try {
      return await firebaseApp.auth().getUserByEmail(email);
    } catch (e) {
      if (e.toString().contains('user-not-found')) {
        return null;
      }
      rethrow;
    }
  }

  Future<bool> isEmailAvailableForUser(String email, String uid) async {
    final existingUser = await getUserByEmailIfExists(email);
    if (existingUser == null) {
      return true;
    }
    return existingUser.uid == uid;
  }
}

File: firebase/functions/src/utils/infra/firebase_auth_utils.ts
import * as admin from 'firebase-admin'
import { firebaseApp } from './firestore_utils'

export class FirebaseAuthUtils {
    async getUser(uid: string): Promise<admin.auth.UserRecord> {
        return firebaseApp.auth().getUser(uid)
    }

    async getUsers(uids: string[]): Promise<admin.auth.UserRecord[]> {
        const promises = uids.map((id) =>
            firebaseApp
                .auth()
                .getUser(id)
                .catch((e) => {
                    console.error(`Failed to get user ${id}:`, e)
                    return null
                })
        )
        const results = await Promise.all(promises)
        return results.filter((r): r is admin.auth.UserRecord => r !== null)
    }
}

export let firebaseAuthUtils = new FirebaseAuthUtils()
export function setFirebaseAuthUtils(instance: FirebaseAuthUtils): void {
    firebaseAuthUtils = instance
}
